### PR TITLE
[Spark] Register API with managed-commits

### PR DESCRIPTION
**Pull request by prakharjain09**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

This PR adds a register API with managed commits.

1. Initial snapshot shouldn't have CommitStore.

2. Introduce concept of managedCommitTableConf. This info is unique to the table. e.g. tableId etc.
- with this change, now configuration has 2 table properties related to managed commits:
2.1 `managedCommitOwnerName` - name of the MC Owner
2.2 `managedCommitOwnerConf` - configuration related to owner
2.3 `managedCommitTableConf` - managed commit configuration related to table

`2.1` and `2.2` are used to initialize CommitStore.
`2.3` is passed as an argument to CommitStore{commit/getCommits/backfillUpto} APIs as it is needed to uniquely identify the table by the owner.

3. CommitStore pre-registration support. If a table is transitioning for FS -> MC, we register with commit-store and get the `managedCommitTableConf` for it. This conf is then set in Metadata.

4. This PR moves the `FS -> MC` commit (or 0th commit) to filesystem as forcing to go it via Commit Owner adds additional complexity and doesn't give any major advantage. So whenever a transition is happening, the commit goes through previous owner:
i.e. For FS -> MC, commit goes through FS. For MC -> FS, commit goes through MC.

5. This PR also introduces a `TableCommitStore` which is a helper class / simplified version of CommitStore. It takes care of passing basic things by itself. e.g. logPath, managedCommitTableMetadata, hadoopConf, logPath.

## How was this patch tested?

UTs

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala [hunk 1]
```diff
@@ -68,15 +68,17 @@ private[delta] case class CurrentTransactionInfo(
    */
   lazy val finalActionsToCommit: Seq[Action] = commitInfo ++: actions
 
-  var newMetadata: Option[Metadata] = None
-  var newProtocol: Option[Protocol] = None
```
Comment (dhruvarya-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:72]: Why remove this/why was this needed in the first place?

File: spark/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala [hunk 2]
```diff
@@ -68,15 +68,17 @@ private[delta] case class CurrentTransactionInfo(
...
+  private var newMetadata: Option[Metadata] = None
 
   actions.foreach {
     case m: Metadata => newMetadata = Some(m)
-    case p: Protocol => newProtocol = Some(p)
     case _ => // do nothing
   }
-  def getUpdateActions(): UpdatedActions = UpdatedActions(commitInfo.get, newMetadata, newProtocol)
+  def getUpdateActions(
```
Comment (dhruvarya-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:77]: [nit] ?
```suggestion
  def getUpdatedActions(
```

File: spark/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala (hunk 1 above)
Comment (prakharjain09) [reply spark/src/main/scala/org/apache/spark/sql/delta/ConflictChecker.scala:72]: previously we were passing only newMetadata and newProtocol when they used to change. Now we pass metadata and protocol (old and new) to commit Store for more flexibility.

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 3]
```diff
@@ -1757,7 +1868,11 @@ trait OptimisticTransactionImpl extends TransactionalWrite
           committed = true
           return (commitVersion, postCommitSnapshot, updatedCurrentTransactionInfo)
         } catch {
-          case _: FileAlreadyExistsException if preCommitCommitStoreOpt.isEmpty =>
+          case _: FileAlreadyExistsException if isFsToMcCommit =>
+            // Don't retry if this commit tries to upgrade the table from filesystem to managed
```
Comment (dhruvarya-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:1872]: Why not retry this?

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 4]
```diff
@@ -1419,6 +1477,57 @@ trait OptimisticTransactionImpl extends TransactionalWrite
...
+          newManagedCommitTableConf = Some(
+            newCommitStore.registerTable(deltaLog.logPath, readVersion, finalMetadata, protocol))
+        case (None, Some(readCommitStore)) =>
+          // MC -> FS conversion
+          val (newOwnerName, newOwnerConf) = CommitOwner.getManagedCommitConfs(snapshot.metadata)
+          logInfo(s"Table ${deltaLog.logPath} transitioning from managed-commit table to " +
+            s"file-system table: [commit owner: $newOwnerName, conf: $newOwnerConf]")
+        case (Some(newCommitStore), Some(readCommitStore))
+            if !readCommitStore.semanticsEquals(newCommitStore) =>
+          // MC1 -> MC2 conversion is not allowed.
```
Comment (dhruvarya-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:1514]: Why are direct transitions between managed commit owners not allowed?

File: spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala [hunk 5]
```diff
@@ -29,7 +29,7 @@ import scala.util.control.NonFatal
 
 import com.databricks.spark.util.TagDefinitions.TAG_ASYNC
 import org.apache.spark.sql.delta.actions.Metadata
-import org.apache.spark.sql.delta.managedcommit.{Commit, CommitStore, GetCommitsResponse}
+import org.apache.spark.sql.delta.managedcommit.{Commit, GetCommitsResponse, TableCommitStore => CommitStore}
```
Comment (dhruvarya-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala:32]: `TableCommitStore => CommitStore` is not a good long-term solution. Can you refactor this in a follow up? Please don't do it in this one because this one is already very big.

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala (hunk 3 above)
Comment (prakharjain09) [reply spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:1872]: The retry would need us to register again with the commit-owner. Currently we register only once in the begiinning on txn start.

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala (hunk 4 above)
Comment (prakharjain09) [reply spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:1514]: The managed-commit spec forbids direct change of commit owners.

File: spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala (hunk 5 above)
Comment (prakharjain09) [reply spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala:32]: Yes makes sense.

**Review by dhruvarya-db — APPROVED** [review]

File: spark/src/main/scala/org/apache/spark/sql/delta/managedcommit/AbstractBatchBackfillingCommitStore.scala [hunk 6]
```diff
@@ -105,15 +120,24 @@ trait AbstractBatchBackfillingCommitStore extends CommitStore with Logging {
     commitPath.getFileSystem(hadoopConf).getFileStatus(commitPath)
   }
 
+  private def isManagedCommitToFSConversion(
+      commitVersion: Long,
+      updatedActions: UpdatedActions): Boolean = {
+    val oldMetadataHasManagedCommits = updatedActions.oldMetadata.managedCommitOwnerName.nonEmpty
+    val newMetadataHasManagedCommits = updatedActions.newMetadata.managedCommitOwnerName.nonEmpty
+    oldMetadataHasManagedCommits && !newMetadataHasManagedCommits && commitVersion > 0
```
[the remark this answers is not in the document]
Comment (prakharjain09) [reply spark/src/main/scala/org/apache/spark/sql/delta/managedcommit/AbstractBatchBackfillingCommitStore.scala:128]: If old owner and new owner both are present, then we compare commitOwners also using semanticEquals to identify they are really the same endpoint underneath. We don't allow MC1 to MC2 conversion and so we error out. This happens in OptimisticTransaction.scala.

Once the control comes here, we have already done those validations. But the commit owner is free to validate stuff again on its own and throw required exceptions.

File: spark/src/main/scala/org/apache/spark/sql/delta/managedcommit/InMemoryCommitStore.scala [hunk 7]
```diff
@@ -126,11 +158,28 @@ class InMemoryCommitStore(val batchSize: Long) extends AbstractBatchBackfillingC
...
+    perTableMap.compute(logPath, (_, existingData) => {
+      if (existingData != null) {
+        if (existingData.lastRatifiedCommitVersion != -1) {
+          throw new IllegalStateException(s"Table $logPath already exists in the commit store.")
+        }
+        // If lastRatifiedCommitVersion is -1 i.e. the commit store has never attempted any commit
+        // for this table => this table was just pre-registered. If there is another newer
+        // pre-registration request, we can take it else we should reject pre-registration request
+        // for older versions.
+        if (existingData.maxCommitVersion >= currentVersion) {
```
[the remark this answers is not in the document]
Comment (prakharjain09) [reply spark/src/main/scala/org/apache/spark/sql/delta/managedcommit/InMemoryCommitStore.scala:176]: Good catch. This should be `currentVersion < existingData.maxCommitVersion`.

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 8]
```diff
@@ -1737,6 +1846,8 @@ trait OptimisticTransactionImpl extends TransactionalWrite
 
     var commitVersion = attemptVersion
     var updatedCurrentTransactionInfo = currentTransactionInfo
+    val isFsToMcCommit =
+      snapshot.metadata.managedCommitOwnerName.isEmpty && metadata.managedCommitOwnerName.nonEmpty
```
[the remark this answers is not in the document]
Comment (prakharjain09) [reply spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:1850]: We allow retries on conflict for downgrade scenario as it doesn't involve pre-registration.

**Review by scovich — APPROVED** [review]
LGTM
